### PR TITLE
ref(ts) Convert SelectControl to typescript

### PR DESCRIPTION
**Pull request by markstory**
We have a few components that wrap this component and having types here will help add types elsewhere. I had to make a number of concessions as the types in react-select use `any` in quite a few places.

### CODE REVIEW

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx [hunk 1]
```diff
@@ -48,13 +57,36 @@ const combineStyles = (newStyles, baseStyles) => {
...
+type LegacyProps = React.ComponentProps<typeof SelectControlLegacy>;
+
+function SelectControl(props: ControlProps) {
   // TODO(epurkhiser): We should remove all SelectControls (and SelectFields,
   // SelectAsyncFields, etc) that are using this prop, before we can remove the
   // v1 react-select component.
   if (props.deprecatedSelectControl) {
     const {deprecatedSelectControl: _, ...legacyProps} = props;
-    return <SelectControlLegacy {...legacyProps} />;
+    return <SelectControlLegacy {...((legacyProps as unknown) as LegacyProps)} />;
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/forms/selectControl.tsx:89]: In parallel I'm trying to make this path go away. I didn't want to put the effort into making types flow nicely here.

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx [hunk 2]
```diff
@@ -48,13 +57,36 @@ const combineStyles = (newStyles, baseStyles) => {
...
+  choices?: Choices | ((props: ControlProps) => Choices);
+  /**
+   * Use react-select v2. Deprecated, don't make more of this.
+   */
+  deprecatedSelectControl?: boolean;
+};
+
+type LegacyProps = React.ComponentProps<typeof SelectControlLegacy>;
+
+function SelectControl(props: ControlProps) {
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/components/forms/selectControl.tsx:83]: I believe ReactSelect lets you use generics for the values. Can we bring that in too?

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx [hunk 3]
```diff
@@ -137,19 +173,19 @@ const SelectControl = props => {
         backgroundColor: theme.active,
       },
     }),
-    valueContainer: provided => ({
+    valueContainer: (provided: React.CSSProperties) => ({
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/components/forms/selectControl.tsx:176]: I wonder if react select has better types for this?

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx [hunk 4]
```diff
@@ -64,23 +96,27 @@ const SelectControl = props => {
...
+      // makes code harder to understand.
       color: theme.formText,
       display: 'flex',
+      // @ts-ignore
       background: theme.background,
+      // @ts-ignore
       border: `1px solid ${theme.border}`,
       borderRadius: theme.borderRadius,
+      // @ts-ignore
       boxShadow: `inset ${theme.dropShadowLight}`,
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/components/forms/selectControl.tsx:120]: Can we put all these ignored props next to each other

maybe we could even do it as 

```tsx
// @ts-ignore
...{
	// Ignored props go in here
}
```

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx [hunk 5]
```diff
@@ -1,29 +1,31 @@
...
 import {IconChevron, IconClose} from 'app/icons';
 import space from 'app/styles/space';
+import {Choices} from 'app/types';
 import convertFromSelect2Choices from 'app/utils/convertFromSelect2Choices';
+import {Theme} from 'app/utils/theme';
 
 import SelectControlLegacy from './selectControlLegacy';
 
-const ClearIndicator = props => (
+const ClearIndicator = (props: any) => (
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/components/forms/selectControl.tsx:16]: Can you just use `React.ComponentProps<typeof selectComponents.ClearIndicator>`

Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/components/forms/selectControl.tsx:16]: Maybe even better would be to declare the variable as 

`React.ComponentType<typeof selectComponents.ClearIndicator>`

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx (hunk 2 above)
Comment (markstory) [reply src/sentry/static/sentry/app/components/forms/selectControl.tsx:83]: I can give it a shot. Hopefully the styled components, HOC and forwarded ref don't interfere.

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx (hunk 3 above)
Comment (markstory) [reply src/sentry/static/sentry/app/components/forms/selectControl.tsx:176]: It [doesn't sadly](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/7914158c56935955528e4f59a1ca2766641dedeb/types/react-select/src/styles.d.ts#L39-L70) at least not in the version we're using. I can try updating the type definitions after getting this merged to see if we can improve these.

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx (hunk 5 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/components/forms/selectControl.tsx:16]: Maybe those types are poor too though 😬

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx [hunk 6]
```diff
@@ -32,12 +34,19 @@ const MultiValueRemove = props => (
...
+ *
+ * @param newStyles the styles to apply on top of base
+ * @param baseStyles the style to override
  */
-const combineStyles = (newStyles, baseStyles) => {
-  return Object.keys(newStyles || {}).reduce((computedStyles, key) => {
-    const styleFunc = (provided, state) =>
+const combineStyles = (
+  newStyles: StylesConfig | undefined,
+  baseStyles: StylesConfig
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/forms/selectControl.tsx:43]:
```suggestion
  baseStyles: StylesConfig
  newStyles?: StylesConfig,
```

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx (hunk 5 above)
Comment (markstory) [reply src/sentry/static/sentry/app/components/forms/selectControl.tsx:16]: Yes. Currently that is a long way to say `any`. :cry:

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx (hunk 6 above)
Comment (markstory) [reply src/sentry/static/sentry/app/components/forms/selectControl.tsx:43]: I'm going to remove this function entirely. I found a similar function in react-select that we can import.

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx (hunk 2 above)
Comment (markstory) [reply src/sentry/static/sentry/app/components/forms/selectControl.tsx:68]: I'm going to revisit adding generics when I update the types package. We currently get basic types for the select options (label, value). While we  have a few select controls that extend that interface, they don't remove from label/value.

File: src/sentry/static/sentry/app/components/forms/selectControl.tsx [hunk 7]
```diff
@@ -292,14 +314,30 @@ const SelectControl = props => {
...
 const SelectControlWithTheme = withTheme(SelectControl);
 
-const SelectPicker = ({async, creatable, forwardedRef, ...props}) => {
+type PickerProps = ControlProps & {
+  /**
+   * Enable async option loading.
+   */
+  async?: boolean;
+  /**
+   * Enable 'create' mode which allows values to be created inline
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/forms/selectControl.tsx:328]:
```suggestion
   * Enable 'create' mode which allows values to be created inline.
```

just to be consistent :)

File: src/sentry/static/sentry/app/components/selectMembers/index.tsx [hunk 8]
```diff
@@ -213,6 +214,8 @@ class SelectMembers extends React.Component<Props, State> {
       return;
     }
 
+    // @ts-ignore The types for react-select don't cover this property.
+    // or the type of selectRef is incorrect.
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/selectMembers/index.tsx:218]:
```suggestion
    // @ts-ignore The types for react-select don't cover this property
    // or the type of selectRef is incorrect.
```

**Review by priscilawebdev — APPROVED** [review]

**Pull request merged**